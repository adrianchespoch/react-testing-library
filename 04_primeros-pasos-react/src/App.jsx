import { useState } from 'react';
import PropTypes from 'prop-types';

import './App.css';
import './assets/tnrbjwasofaag8cm7l9v.jpg';
import './assets/react.svg';

function App({ title, initCounter }) {
  const [counter, setCounter] = useState(initCounter);

  // console.log('Render TODO el FC');  // Cuando CAMBIA el STATE se vuelve a Ejecutar el FC

  // // Si la f(x) esta fuera del FC no se reasisgna espacio en memoria con c/rerender
  const handleCounter = () => {
    // setCounter(counter + 1);
    setCounter(c => c + 1);
  };
  const handleSubtract = () => setCounter(counter - 1);
  const handleReset = () => setCounter(initCounter);

  return (
    <div className="App">
      <h1 data-testid="test-title">{title}</h1>
      <h2>{counter}</h2>

      {/* <Counter /> */}
      <div className="card">
        <button className="btn-increase" onClick={handleCounter}>
          +1
        </button>
        <button id="btn-decrease" onClick={handleSubtract}>
          -1
        </button>
        <button aria-label="btn-reset" onClick={handleReset}>
          Reset
        </button>
      </div>

      <p>
        Edit <code>src/App.jsx</code> and save to test HMR
      </p>
      <p className="read-the-docs">
        Click on the Vite and React logos to learn more
      </p>
    </div>
  );
}

export default App;

// // // PropTypes
App.propTypes = {
  title: PropTypes.string.isRequired,
  initCounter: PropTypes.number.isRequired,
};

// // DefaultProps: Entran antes q los PropTypes, evitan error si es  isRequired
App.defaultProps = {
  title: 'No title!',
};
