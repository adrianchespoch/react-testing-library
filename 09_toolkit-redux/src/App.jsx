import { useDispatch, useSelector } from 'react-redux';

import { decrement, increment, incrementByAmount } from './store/slices/counter';
import './App.css';

function App() {
  // Extraemos lo q definimos en el initState del slice
  const { counter } = useSelector(state => state.counter);

  // ya sabe a q reducer va a llegar
  const dispatch = useDispatch();

  return (
    <div className="App">
      <h1>Count is: {counter}</h1>

      <div className="card">
        <button onClick={() => dispatch(increment())}>Increment</button>
        <button onClick={() => dispatch(decrement())}>Decrement</button>
        <button onClick={() => dispatch(incrementByAmount(3))}>Increment by 2</button>
      </div>
    </div>
  );
}

export default App;
