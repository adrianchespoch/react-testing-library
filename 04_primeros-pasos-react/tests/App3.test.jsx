import { fireEvent, render, screen } from '@testing-library/react';
import App from '../src/App';

describe('Pruebas sobre el counter', () => {
  const title = 'React';
  const initCounter = 10;

  test('should match con el snapshot', () => {
    const { container } = render(
      <App title={title} initCounter={initCounter} />
    );

    expect(container).toMatchSnapshot();
  });

  test('should return 100', () => {
    render(<App title={title} initCounter={initCounter} />);

    expect(screen.getByText(initCounter)).toBeTruthy();
    expect(screen.getByRole('heading', { level: 2 }).innerHTML).toContain(
      String(initCounter)
    );
  });

  test('should increase with +1 btn', () => {
    render(<App title={title} initCounter={initCounter} />);

    // eventos
    fireEvent.click(screen.getByText('+1'));

    expect(screen.getByText('11')).toBeTruthy();
  });

  test('should decrementar con el btn -1', () => {
    render(<App title={title} initCounter={initCounter} />);

    fireEvent.click(screen.getByText('-1'));
    // screen.debug();

    expect(screen.getByText('9')).toBeTruthy();
  });

  test('should reset', () => {
    render(<App title={title} initCounter={initCounter} />);

    fireEvent.click(screen.getByText('+1'));
    fireEvent.click(screen.getByText('+1'));
    fireEvent.click(screen.getByText('+1'));

    fireEvent.click(screen.getByText('-1'));
    // screen.debug();

    // // // aria-label="btn-reset"   <-  HTML
    // fireEvent.click(screen.getByText('Reset'));
    fireEvent.click(screen.getByRole('button', { name: 'btn-reset' }));

    expect(screen.getByText('10')).toBeTruthy();
  });
});
