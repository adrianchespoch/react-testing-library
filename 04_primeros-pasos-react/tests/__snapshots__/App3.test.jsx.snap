// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el counter should match con el snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <h1
      data-testid="test-title"
    >
      React
    </h1>
    <h2>
      10
    </h2>
    <div
      class="card"
    >
      <button
        class="btn-increase"
      >
        +1
      </button>
      <button
        id="btn-decrease"
      >
        -1
      </button>
      <button
        aria-label="btn-reset"
      >
        Reset
      </button>
    </div>
    <p>
      Edit 
      <code>
        src/App.jsx
      </code>
       and save to test HMR
    </p>
    <p
      class="read-the-docs"
    >
      Click on the Vite and React logos to learn more
    </p>
  </div>
</div>
`;
